### DISCUSSION TITLE
Allow running with different RUSTFLAGS and take "max" of all runs

### DISCUSSION THREAD

**Discussion by wmmc88** [obi1kenobi/cargo-semver-checks#1012] [Ideas]
Some crates use `cfg`s to define non-additive features in order to not break [Cargo's assumption that features are additive](https://doc.rust-lang.org/cargo/reference/features.html#feature-unification). Some examples include [tokio](https://github.com/tokio-rs/tokio/blob/f8948ea021bd4b5626bbd8dee89bee51ba766326/tokio/Cargo.toml#L103), [uuid](https://github.com/uuid-rs/uuid/blob/4473398413f7e3f4103374559b8d038b4801329a/src/lib.rs#L225), [windows-rs](https://github.com/microsoft/windows-rs/blob/30a574a122d2d03dff6203f08c81a284114ecb88/.cargo/config.toml#L5).

What if `cargo-semver-checks` could run with each of the cfg flags individually, and take the "max" version of all the runs?

**Comment by obi1kenobi** [comment]
Hi! Thanks for the suggestion.

Could you say more about how such a feature would be useful to users? I'm having a bit of trouble visualizing it — I've never used `cfg` clauses in this way, and two of the three examples you provided seem to be around explicitly-unstable features where I'm not sure cargo-semver-checks would be that valuable.

**Reply by wmmc88** [reply]
Here is a relevant discussion on a highly popular repo on why they switched from using non-additive cargo features to using global `cfg` options for non-additive features: https://github.com/RustCrypto/block-ciphers/issues/245.
 
In practice, these are enabled by setting `RUSTFLAGS` either in `config.toml` or via the env variable: https://docs.rs/aes/latest/aes/#configuration-flags
 
> Could you say more about how such a feature would be useful to users?
 
This non-cargo-feature mechanism allows you to conditionally compile code. A new crate version could remove functions that were only present when `cfg(A)` was specified (or similarly the function may only be deprecated/removed for `cfg(not(B))`. These semver-breaking changes could be caught if you were able to run `cargo-semver-checks` multiple times with different `--cfg` flags enabled, and then take the highest version suggested by the multiple runs.
 
Example:
 
1. A crate has `--cfg A` and `--cfg B` mutually-incompatible flags. Currently, the crate is version `1.2.3`. It currently also has a function that is conditionally compiled:
```rust
#[cfg(A)]
pub fn foo() { 
    println!("A!");
}
```
2. Three changes were introduced:
    1.  `foo` was removed
    2. A new api, `bar`, was added:
    ```rust
    #[cfg(B)]
    pub fn bar() { 
        println!("B");
    }
    ```
    3. Some dependency that was only used internally was updated
 
3. Currently, the behavior would be:
    1.  when no RUSTFLAGS are set, cargo-semver-checks would suggest new version of `1.2.4` (patch bump from dependency update)
    2. If `RUSTFLAGS="--cfg A"`,  cargo-semver-checks would suggest new version of `2.0.0` (major bump from removal of a public item)
    3. If `RUSTFLAGS="--cfg B"`,  cargo-semver-checks would suggest new version of `1.3.0` (minor bump from addition of new public item)
    
Ideally, `cargo-semver-checks` could be configured to run these 3 runs, and then decide that the proper new version should be `2.0.0`

**Reply by obi1kenobi** [reply]
Would the outcome be mostly the same if the user were to invoke `cargo-semver-checks` three times with different `--cfg` flags?

Asking because a system for configuring multiple c-s-c runs in one invocation is obviously going to require more work than just exposing the ability to supply a different `--cfg` option for each run separately.

**Reply by wmmc88** [reply]
I think running it 3 times with separate cfg flags would be similar, but:
* it would require parsing out the new suggested version from the output 3 times (since csc has no option for some machine-readable output related #56, #6 )
* it would require an external script to take the min new version output of the 3 runs, parse the semver versions, and select the highest one

AFAIU, csc only uses the output of `rustdoc`(and not `rustc`), so setting `RUSTDOCFLAGS="--cfg A"` and similar should work: https://doc.rust-lang.org/rustdoc/command-line-arguments.html?highlight=cfg#--cfg-passing-configuration-flags, [except that `RUSTDOCFLAGS` gets overridden currently](https://github.com/obi1kenobi/cargo-semver-checks/blob/a80eb2e78ad7e6d3dd4ff60cac62e0e54006e786/src/data_generation/generate.rs#L269-L270). If this were changed to respect existing `RUSTDOCFLAGS`, then its conceivable that someone could just write their own binary using the csc library to run N times and find the minimum version floor for a new release.

**Reply by obi1kenobi** [reply]
As of the upcoming release, we actually also use `cargo metadata` so we can lint package manifests. The announcement for that will go out shortly, so there's no way you could have known this though :)

(EDIT: As far as I can tell, `RUSTFLAGS` and `RUSTDOCFLAGS` don't affect `cargo metadata`, so it's just rustdoc that we need to tell about `--cfg A` style settings.)

#1016 makes it so that `RUSTDOCFLAGS` is preserved. It also includes a test case to prove that the new code path works with `--cfg A` syntax. Thanks for the suggestion on that!

I think building the infra for making multiple runs would be interesting, and would also help for scanning different feature configurations to find non-additive features! Unfortunately, it would take quite a bit of design, implementation, and testing work.

Due to the scope, it's not something I can take on right now as part of my regular work on this project. If this functionality would be useful to your team, I'd love to discuss developing it as part of a commercial development and support contract. Is that something you or your team would be open to?

**Reply by wmmc88** [reply]
Thanks for the fix for flag override behavior! I think at this time, we're going to prototype some things using the existing functionalities

**Reply by obi1kenobi** [reply]
Neat! I'd love to know what you figure out — if there's a system you build and find pleasant to use, I'd love to take a look and possibly even a PR.

**Reply by obi1kenobi** [reply]
The v0.37 release is currently being built btw, it should be out within a few minutes: https://github.com/obi1kenobi/cargo-semver-checks/actions/runs/12172704503
